trace-obfuscation/src/replacer.rs

```diff
@@ -0,0 +1,198 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present
+// Datadog, Inc.
+
+use crate::pb;
+use regex::Regex;
+
+pub trait TraceTagReplacer {
+    fn replace_trace_tags(trace: &mut [pb::Span], rules: &[ReplaceRule]);
+}
+
+#[derive(Debug)]
+pub struct ReplaceRule<'a> {
+    // name specifies the name of the tag that the replace rule addresses. However,
+    // some exceptions apply such as:
+    // • "resource.name" will target the resource
+    // • "*" will target all tags and the resource
+    name: &'a str,
+
+    // re holds the regex pattern for matching.
+    re: regex::Regex,
+
+    // repl specifies the replacement string to be used when Pattern matches.
+    repl: &'a str,
+}
+
+struct DefaultTraceTagReplacer {}
+
+impl TraceTagReplacer for DefaultTraceTagReplacer {
+    /// replace_trace_tags replaces the tag values of all spans within a trace with a given set of rules.
+    fn replace_trace_tags(trace: &mut [pb::Span], rules: &[ReplaceRule]) {
+        for rule in rules {
+            for span in &mut *trace {
+                match rule.name {
+                    "*" => {
+                        for (_, val) in span.meta.iter_mut() {
+                            *val = rule.re.replace_all(val, rule.repl).to_string();
+                        }
+                    }
+                    "resource.name" => {
+                        span.resource = rule.re.replace_all(&span.resource, rule.repl).to_string();
+                    }
+                    _ => {
+                        if let Some(val) = span.meta.get_mut(rule.name) {
+                            let replaced_tag = rule.re.replace_all(val, rule.repl).to_string();
+                            *val = replaced_tag;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// parse_rules_from_string takes an array of rules, represented as an array of length 3 arrays
+/// holding the tag name, regex pattern, and replacement string as strings.
+/// * returns a vec of ReplaceRules
+pub fn parse_rules_from_string<'a>(
```

Comment:
this may be changed in the future depending on how the "agentless" agent consumes config variables, and how we want to parse the "replace_tags" config variables.